### PR TITLE
feature(sunjx): add high entropy token selection

### PR DESCRIPTION
**Pull request by Jiaxuan-Sun**

### CODE REVIEW

File: lightrft/models/actor_language.py [hunk 1]
```diff
@@ -288,20 +290,43 @@ def forward(
             return output
 
         log_probs = log_probs_from_logits(output["logits"][:, :-1, :], sequences[:, 1:])
+        
+        # Calculate entropy for action tokens (for high-entropy token identification)
+        # Only compute entropy when high_entropy_token_ratio is not 0
+        if self.high_entropy_token_ratio > 0.0:
+            action_logits = output["logits"][:, :-1, :]  # Shape: (batch, seq_len-1, vocab_size)
+            action_entropy = entropy_from_logits(action_logits)  # Shape: (batch, seq_len-1)
```
Comment (puyuan1996) [inline lightrft/models/actor_language.py]: Why :-1？

File: lightrft/trainer/fast_exp_maker.py [hunk 2]
```diff
@@ -1794,8 +1812,10 @@ def _pack_experience(
                 info,
                 kl,
             )
+            exp.action_entropy = output.action_entropy
```
Comment (puyuan1996) [inline lightrft/trainer/fast_exp_maker.py]: Is it possible to include action_entropy in the ExperienceVL and Experience definitions?

File: render_trajectories.html [hunk 3]
```diff
@@ -0,0 +1,532 @@
+<!DOCTYPE html>
```
Comment (PaParaZz1) [inline render_trajectories.html]: we should move this visualization to `examples/entropy_viz` directory.

File: lightrft/trainer/spmd_ppo_trainer.py [hunk 4]
```diff
@@ -210,8 +213,69 @@ def ppo_train(self, global_steps=0):  # Currently using this rewritten ppo_train
                     experience = make_experience_batch(items, packing_samples=self.packing_samples)
                 experience.to_device(device)
 
+                # Create entropy_mask if high_entropy_token_ratio > 0 and action_entropy is available
+                entropy_mask = None
+                if hasattr(experience, 'action_entropy') and experience.action_entropy is not None:
+                    if hasattr(self.actor, 'high_entropy_token_ratio') and self.actor.high_entropy_token_ratio > 0.0:
+                        from lightrft.models.utils import create_high_entropy_mask
```
Comment (PaParaZz1) [inline lightrft/trainer/spmd_ppo_trainer.py]: move import to the top side

File: lightrft/trainer/fast_exp_maker.py (hunk 2 above)
Comment (PaParaZz1) [reply lightrft/trainer/fast_exp_maker.py]: +1, we should enable it in the dataclass definition and set the default value to None

Comment (PaParaZz1) [reply lightrft/trainer/fast_exp_maker.py]: and the below creation can also include action_entropy, we don't need extra assignment code

File: lightrft/models/loss.py [hunk 5]
```diff
@@ -64,36 +84,180 @@ def forward(self, logits: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
...
+        use_cpg_loss: bool = False,
+        high_entropy_token_ratio: float = 0.0,
+        entropy_mask: Optional[torch.Tensor] = None,
+    ) -> None:
         super().__init__()
         self.clip_eps = clip_eps
         self.use_dapo = use_dapo
         self.use_cpg_loss = use_cpg_loss
+        self.high_entropy_token_ratio = high_entropy_token_ratio
+        self.entropy_mask = entropy_mask
```
Comment (PaParaZz1) [inline lightrft/models/loss.py]: why do we need entropy_mask in init function? Do we have the default settings for mask?

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: lightrft/models/loss.py [hunk 6]
```diff
@@ -64,36 +84,180 @@ def forward(self, logits: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
...
+            final_mask = action_mask
+
         if self.use_cpg_loss:
             clipped_log_probs = torch.where(
                 advantages > 0, torch.clamp(log_probs, max=torch.log(torch.tensor(1 + self.clip_eps)) + old_log_probs),
                 torch.clamp(log_probs, min=torch.log(torch.tensor(1 - self.clip_eps)) + old_log_probs)
             )
             loss = -clipped_log_probs * advantages
-            loss = (loss * action_mask).sum() / action_mask.sum()
+            loss = masked_mean(loss, final_mask, dim=-1).mean()
```
[the remark this answers is not in the document]
Comment (Jiaxuan-Sun) [reply lightrft/models/loss.py]: maintain consistency with the PPO Loss in the following eight lines

File: view_trajectories.sh [hunk 7]
```diff
@@ -0,0 +1,56 @@
+#!/bin/bash
+# 简单的可视化启动脚本
```
Comment (puyuan1996) [inline view_trajectories.sh]: 这个是错误上传上来的吗

File: lightrft/utils/trajectory_saver.py [hunk 8]
```diff
@@ -35,18 +35,34 @@ class TrajectorySaver:
...
     def __init__(
         self,
         save_dir: str,
         tokenizer: Any,
         save_images_separately: bool = True,
         max_image_size: int = 512,
+        mark_high_entropy_tokens: bool = False,
+        high_entropy_token_ratio: float = 0.2,
+        high_entropy_marker_start: str = "<HIGH_ENTROPY>",
+        high_entropy_marker_end: str = "</HIGH_ENTROPY>",
```
Comment (puyuan1996) [inline lightrft/utils/trajectory_saver.py]: 改成新的通过html渲染标记方式后，这个arg还在使用吗 如果没有就删除了吧

File: lightrft/trainer/spmd_ppo_trainer.py [hunk 9]
```diff
@@ -210,8 +214,68 @@ def ppo_train(self, global_steps=0):  # Currently using this rewritten ppo_train
...
+                            # Align action_entropy and action_mask shapes
+                            if action_mask is not None:
+                                if len(action_mask.shape) != 2:
+                                    raise ValueError(
+                                        f"action_mask must be 2D tensor (batch_size, num_actions), "
+                                        f"got shape {action_mask.shape}"
+                                    )
+                                
+                                # If shapes don't match, align them to the minimum length
+                                if action_mask.shape != action_entropy.shape:
```
Comment (puyuan1996) [inline lightrft/trainer/spmd_ppo_trainer.py]: 这里确实有这么多种情况吗 是否可以简化

File: lightrft/models/loss.py [hunk 10]
```diff
@@ -64,36 +84,174 @@ def forward(self, logits: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
...
+    for efficient training.
+
+    **Supported Algorithms:**
+
+    - **PPO (Proximal Policy Optimization)**: Default mode using standard clipped surrogate
+      objective. The loss is computed as ``-min(ratio * advantages, clipped_ratio * advantages)``
+      where ``ratio = exp(log_probs - old_log_probs)`` and clipping is applied to prevent
+      large policy updates.
+
+    - **CPGD (Clipped Policy Gradient Descent)**: Enabled via ``use_cpg_loss=True``. Uses
```
Comment (puyuan1996) [inline lightrft/models/loss.py]: Clipped Policy Gradient Optimization with Policy Drift (CPGD)
